### PR TITLE
refactor(core): store pending ops per realm

### PR DESCRIPTION
**Pull request by bartlomieju**
Ref https://github.com/denoland/deno/issues/19020#issuecomment-1539959897

This implementation is not complete - when the realm is garbage collected its "pending_ops" container is not removed. CC @andreubotella any pointers here?

### CODE REVIEW

**Comment by bartlomieju** [comment]
> This implementation is not complete - when the realm is garbage collected its "pending_ops" container is not removed. CC @andreubotella any pointers here?

A work around for this problem would be to store `pending_ops` container on the `ContextState`, however we can't do that, because we don't have access to a `scope` in Fast Calls API.

File: core/runtime.rs [hunk 1]
```diff
@@ -180,6 +184,57 @@ pub struct JsRuntimeState {
...
+    let JsRealmState::Active(_, ops) = self.known_realms.get_mut(realm_idx).unwrap() else {
+      unreachable!()
+    };
+    ops
+  }
+
+  pub(crate) fn context<'s>(&self, realm_idx: usize, scope: &mut v8::HandleScope<'s, ()>) -> Option<v8::Local<'s, v8::Context>> {
+    let JsRealmState::Active(weak, _) = self.known_realms.get(realm_idx).unwrap() else {
+      return None;
+    };
```
Comment (bartlomieju) [inline core/runtime.rs:205]: Why can we return `None` here? Isn't it a coding error if the realm is disposed to call this function?

File: core/runtime.rs [hunk 2]
```diff
@@ -180,6 +184,57 @@ pub struct JsRuntimeState {
...
+    weak.to_local(scope)
+  }
+
+  pub(crate) fn context_global(&self, realm_idx: usize, isolate: &mut v8::OwnedIsolate) -> Option<v8::Global<v8::Context>> {
```
Comment (bartlomieju) [inline core/runtime.rs:209]: Ditto

File: core/runtime.rs [hunk 3]
```diff
@@ -180,6 +184,57 @@ pub struct JsRuntimeState {
...
+    let JsRealmState::Active(weak, _) = self.known_realms.get(realm_idx).unwrap() else {
+      return None;
+    };
+    weak.to_global(isolate)
+  }
+
+  pub(crate) fn destroy_realms(&mut self, isolate: &mut v8::Isolate) {
```
Comment (bartlomieju) [inline core/runtime.rs:216]:
```suggestion
  pub(crate) fn destroy_all_realms(&mut self, isolate: &mut v8::Isolate) {
```

File: core/runtime.rs [hunk 4]
```diff
@@ -490,7 +544,7 @@ impl JsRuntime {
       state.inspector = inspector;
       state
         .known_realms
-        .push(v8::Weak::new(&mut isolate, &global_context));
+        .push(JsRealmState::Active(v8::Weak::new(&mut isolate, &global_context), FuturesUnordered::new()));
```
Comment (bartlomieju) [inline core/runtime.rs:547]: Consider adding `JsRealmState::new()` to hide some of the details here

File: core/runtime.rs [hunk 5]
```diff
@@ -617,12 +671,14 @@ impl JsRuntime {
...
-      self
-        .state
-        .borrow_mut()
-        .known_realms
-        .push(v8::Weak::new(scope, context));
+      {
+        let state_rc = self.state.clone();
+        let mut state = self.state.borrow_mut();
+        state.known_realms.push(JsRealmState::Active(v8::Weak::with_guaranteed_finalizer(scope, context, Box::new(move || {
+          // Clear the pending ops from this realm
```
Comment (bartlomieju) [inline core/runtime.rs:678]: Is it a TODO?

**Review by bartlomieju — COMMENTED** [review]
Looking good! Exciting change

File: core/runtime.rs (hunk 1 above)
Comment (mmastrac) [reply core/runtime.rs:205]: This is where it gets complex. As realms exist right now, they are only ever created and never destroyed.

I think I have to clean this code up quite a bit -- if a realm is disposed, it doesn't actually clean up the associated runtime data, just clears the slots in its context which triggers the GC where we can actually clean things up. This means that there's a period of time where we have a realm that's half disposed.

I'm starting to think we should avoid realm indexes and just use Rc<something> instead. Let me try a different approach as I'm not terribly happy with this.

Comment (bartlomieju) [reply core/runtime.rs:205]: Ah I see, that makes sense. Feel free to address it in a followup if you prefer.

**Review by mmastrac — APPROVED** [review]
Co-approving w/ @bartlomieju

**Pull request merged**